crates/bevy_reflect/src/map.rs

```diff
@@ -226,20 +230,38 @@ impl DynamicMap {
     pub fn insert<K: PartialReflect, V: PartialReflect>(&mut self, key: K, value: V) {
         self.insert_boxed(Box::new(key), Box::new(value));
     }
+
+    fn internal_hash(value: &dyn PartialReflect) -> u64 {
+        value.reflect_hash().expect(hash_error!(value))
+    }
+
+    fn internal_eq<'a>(
+        value: &'a dyn PartialReflect,
+        values: &'a [(Box<dyn PartialReflect>, Box<dyn PartialReflect>)],
+    ) -> impl FnMut(&usize) -> bool + 'a {
+        |&index| {
+            value
+            .reflect_partial_eq(&*values[index].0)
+            .expect("Underlying type does not reflect `PartialEq` and hence doesn't support equality checks")
+        }
+    }
 }
 
 impl Map for DynamicMap {
     fn get(&self, key: &dyn PartialReflect) -> Option<&dyn PartialReflect> {
+        let hash = Self::internal_hash(key);
+        let eq = Self::internal_eq(key, &self.values);
         self.indices
-            .get(&key.reflect_hash().expect(hash_error!(key)))
-            .map(|index| &*self.values.get(*index).unwrap().1)
+            .find(hash, eq)
+            .map(|&index| &*self.values[index].1)
```

Comment:
This is cool, I didn't know you could do this. This works by pattern matching?

Reply:
Yes, it's pattern matching against the reference and binding `index` to the referenced value.